src/Storages/QueryCache.h

```diff
@@ -0,0 +1,221 @@
+    {
+        return ast->getTreeHash() == other.ast->getTreeHash()
+               && header.getNamesAndTypesList() == other.header.getNamesAndTypesList()
+               && settings == other.settings
+               && username == other.username;
+    }
+
+    ASTPtr ast;
+    Block header;
+    Settings settings;
+    std::optional<String> username;
+};
+
+struct CacheKeyHasher
+{
+    size_t operator()(const CacheKey & key) const
+    {
+        auto ast_info = key.ast->getTreeHash();
+        auto header_info = std::hash<String>{}(key.header.getNamesAndTypesList().toString());
+        auto settings_info = settingsHash(key.settings);
+        auto username_info = std::hash<std::optional<String>>{}(key.username);
+
+        return ast_info.first + ast_info.second * 9273
+               + header_info * 9273 * 9273
+               + settings_info * 9273 * 9273 * 9273
+               + username_info * 9273 * 9273 * 9273 * 9273;
+    }
+
+    private:
+        static size_t settingsHash(const Settings & settings)
+        {
+            size_t hash = 0;
+            size_t coefficient = 1;
+            for (const auto & setting : settings)
+            {
+                hash += std::hash<String>{}(setting.getValueString()) * coefficient;
+                coefficient *= 53;
+            }
+            return hash;
+        }
+};
+
+struct QueryWeightFunction
+{
+    size_t operator()(const Data & data) const
+    {
+        const Block & block = data.first;
+        const Chunks & chunks = data.second;
+
+        size_t res = 0;
+        for (const auto & chunk : chunks)
+        {
+            res += chunk.allocatedBytes();
+        }
+        res += block.allocatedBytes();
+
+        return res;
+    }
+};
+
+class CacheRemovalScheduler
+{
+private:
+    using timestamp = std::chrono::time_point<std::chrono::high_resolution_clock>;
+    using duration = std::chrono::high_resolution_clock::duration;
+public:
+    void scheduleRemoval(duration duration, CacheKey cache_key)
+    {
+        std::lock_guard lock(mutex);
+        TimedCacheKey timer = {now() + duration, cache_key};
+        queue.push(timer);
+        if (queue.top() == timer)
+        {
+            timer_cv.notify_one();
+        }
+    }
+
+    template <typename Cache>
+    [[noreturn]] void processRemovalQueue(Cache * query_cache)
+    {
+        while (true)
+        {
+            std::unique_lock lock(mutex);
+            const std::optional<TimedCacheKey> awaited_timer = nextTimer();
+
+            timer_cv.wait_until(lock,
+                                awaited_timer.has_value() ? awaited_timer->time : infinite_time,
+                                [&]() { return awaited_timer != nextTimer() || (awaited_timer.has_value() && awaited_timer->time <= now()); }
+            );
+
+            if (awaited_timer.has_value() && awaited_timer->time <= now())
+            {
+                query_cache->remove(awaited_timer->cache_key);
```

Comment:
This should be done without mutex lock. Because we block threads who try to schedule removal.

Reply:
Fixed this.